core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java

```diff
@@ -1917,27 +1927,68 @@ public Locations locate(String tableName, Collection<Range> ranges)
       rangeList = new ArrayList<>(ranges);
     }
 
-    Map<String,Map<KeyExtent,List<Range>>> binnedRanges = new HashMap<>();
-
+    ClientTabletCache locator = ClientTabletCache.getInstance(context, tableId);
     locator.invalidateCache();
 
     Retry retry = Retry.builder().infiniteRetries().retryAfter(100, MILLISECONDS)
         .incrementBy(100, MILLISECONDS).maxWait(2, SECONDS).backOffFactor(1.5)
         .logInterval(3, MINUTES).createRetry();
 
-    // ELASTICITY_TODO this will cause tablets to be hosted, but that may not be desired
-    while (!locator.binRanges(context, rangeList, binnedRanges).isEmpty()) {
-      context.requireTableExists(tableId, tableName);
-      context.requireNotOffline(tableId, tableName);
-      binnedRanges.clear();
-      try {
-        retry.waitForNextAttempt(log,
-            String.format("locating tablets in table %s(%s) for %d ranges", tableName, tableId,
-                rangeList.size()));
-      } catch (InterruptedException e) {
-        throw new RuntimeException(e);
+    ArrayList<KeyExtent> locationLess = new ArrayList<>();
+    Map<String,Map<KeyExtent,List<Range>>> binnedRanges = new HashMap<>();
+
+    BiConsumer<CachedTablet,Range> rangeConsumer = (cachedTablet, range) -> {
+      // We want tablets that are currently hosted (location present) and
+      // are where their tablet hosting goal is ALWAYS (not OnDemand)
+      if (cachedTablet.getTserverLocation().isPresent()
+          && cachedTablet.getGoal() == TabletHostingGoal.ALWAYS) {
+        ClientTabletCacheImpl.addRange(binnedRanges, cachedTablet, range);
+      } else {
+        locationLess.add(cachedTablet.getExtent());
       }
-      locator.invalidateCache();
+    };
+
+    try {
+
+      List<Range> failed =
+          locator.findTablets(context, rangeList, rangeConsumer, LocationNeed.NOT_REQUIRED);
+
+      while (!failed.isEmpty() || !locationLess.isEmpty()) {
+
+        log.warn("failures: {}", failed);
+        log.warn("w/out location: {}", locationLess);
+        context.requireTableExists(tableId, tableName);
+        context.requireNotOffline(tableId, tableName);
+
+        boolean allAlwaysHosted = context.getAmple().readTablets().forTablets(locationLess)
+            .fetch(PREV_ROW, HOSTING_GOAL).build().stream()
+            .allMatch(tmeta -> tmeta.getHostingGoal() == TabletHostingGoal.ALWAYS);
+
+        if (!allAlwaysHosted) {
+          throw new AccumuloException(
+              "TableOperations.locate() only works with tablets that have a hosting goal of "
+                  + TabletHostingGoal.ALWAYS
+                  + ". Tablets with other hosting goals were seen.  table:" + tableName
+                  + " table id:" + tableId);
+        }
```

Comment:
Since we are clearing the cache before the goals in the cache should be really fresh and would could probably drop the call to ample only use the data from the cache. In the rangeConsumer passed to the cache we could add to a list of extents (or set an atomic boolean to true) that do not have a goal of always.  We can check that and if its not as expected throw this exception.

Reply:
You are saying that we can remove the entire retry loop?

Reply:
Used AtomicBoolean in consumer to identify when tablet found in Range with goal != ALWAYS. Removed call to Ample in 8066fd0.